Enable Serial Greentea FPGA CI Test Shield tests

fkjagodzinski opened the pull request:
This patch builds on top of #32 and brings:

- `tests/mbed_hal_fpga_ci_test_shield/uart` -- Serial greentea test executed with the FPGA CI Test Shield.
- A static pinmap with fixed `MSTD_CONSTEXPR_FN_14` macro.

hugueskamba commented:
Can you please also include this patch?

It helps with ensuring that devices which do not `DEVICE_SERIAL_FC` but do define `DEVICE_SERIAL` compile without error and reduced warnings.
[fpga_shield_uart.patch.txt](https://github.com/MCU-Driver-HAL/MCU-Driver-HAL/files/6842749/fpga_shield_uart.patch.txt)

fkjagodzinski commented:
Fixed the include directive for `DEVICE_SERIAL_FC=0` case and also the unused variable warning.

fkjagodzinski commented on CMakeLists.txt:
```diff
@@ -158,13 +158,13 @@ add_subdirectory(hal)
 # Examples
 add_subdirectory(examples/flash EXCLUDE_FROM_ALL)
 add_subdirectory(examples/gpio EXCLUDE_FROM_ALL)
+add_subdirectory(examples/pwm EXCLUDE_FROM_ALL)
```
OK, removed the unrelated clean-up.

fkjagodzinski commented:
Commits rebased and squashed. This is ready for review and merge. :wink:

ithinuel commented on tests/mbed_hal_fpga_ci_test_shield/uart/main.cpp:
```diff
@@ -209,11 +209,9 @@ static void uart_test_common(int baudrate, int data_bits, SerialParity parity, i
-            // Wait (no longer than twice the time of one packet transfer) for
-            // the FPGA to receive data and update the byte counter.
-        }
+        // Wait twice the time of one packet transfer for
```
why this change in behaviour when `us_ticker_read` is available?

fkjagodzinski replied:
This is simpler. We only want to check if the char has been received by the test shield in a defined time window. Using `us_ticker_read()` also means ticks->time conversion. I don't like all this extra code here; `wait_us()` is sufficient. I'm also aware this means the test takes longer to complete, but I think we can disregard that.

ithinuel replied:
`wait_us` is deceptive. It hides this code and makes the requirement for the us_ticker.
The verbosity of the previous code isn't satisfying but it is better than `wait_us`.

`wait_us` may not use us_ticker on some platform which make this even more complex and is a pitfall we already encountered in the past weeks.

fkjagodzinski replied:
OK, I'll update that not to use `wait_us()`.

fkjagodzinski replied:
Done, 8dc66d008dffd11a79e1d45ba4c74c4d783bcc6c.

ithinuel commented on tests/mbed_hal_fpga_ci_test_shield/uart/main.cpp:
```diff
@@ -410,14 +406,16 @@ Case cases[] = {
     GREENTEA_SETUP(240, "default_auto");
-    srand((unsigned) ticker_read_us(get_us_ticker_data()));
+    us_ticker_init();
+    srand((unsigned) us_ticker_read());
```
This is deceptive. Init is fairly constant so this seed is not random.

fkjagodzinski replied:
True; this as bad as what we had before. Anyway, any idea how to improve that?

ithinuel replied:
Not to use random if it is not random.
Define a constant and get rid of occurences of `rand` is the simplest path forward.

ithinuel commented on tests/mbed_hal_fpga_ci_test_shield/uart/uart_fpga_test.h:
```diff
@@ -22,6 +22,9 @@
 
 #if DEVICE_SERIAL
```
why isn't this occurrence removed as well?

fkjagodzinski replied:
Could you elaborate? I did not remove any `DEVICE_SERIAL`.

ithinuel replied:
I meat, with the other recent PR that did remove their related `DEVICE_*`.
If my understanding is correct, in the grand scheme of things, dependencies on peripherals shall be evaluated by CMake and not done in code with `#ifdef`s.

ithinuel commented on tools/fpga_ci_test_shield/include/fpga_ci_test_shield/test_utils.h:
```diff
@@ -581,9 +581,9 @@ const char *const AnalogoutMaps::name = ANALOGOUT_NAME;
 
-#if DEVICE_SERIAL && 0
-#if DEVICE_SERIAL_FC
+#if DEVICE_SERIAL
```
why are they splitted?

fkjagodzinski replied:
There are cases where `DEVICE_SERIAL` is defined and `DEVICE_SERIAL_FC` is not. The header has to be there also for these cases.

ithinuel commented on hal/include/hal/static_pinmap.h:
```diff
@@ -31,8 +31,32 @@
 #include "qspi_api.h"
 #include "ospi_api.h"
 #include "can_api.h"
+
```
can we not add them and remove the include for that c++ header?

fkjagodzinski replied:
I don't follow. Which parts do you want removed?

ithinuel replied:
line 34 to 59 :)

fkjagodzinski replied:
There are [test cases](https://github.com/MCU-Driver-HAL/MCU-Driver-HAL/blob/efab67e6bbe6120d2e0418f16698f21bbd1782dd/tests/mbed_hal_fpga_ci_test_shield/uart/main.cpp#L354) that require [`get_uart_pinmap()`](https://github.com/MCU-Driver-HAL/MCU-Driver-HAL/blob/efab67e6bbe6120d2e0418f16698f21bbd1782dd/tests/mbed_hal_fpga_ci_test_shield/uart/main.cpp#L142). This is why we need to `#include "hal/static_pinmap.h"`. This however, declares all it's functions as `MSTD_CONSTEXPR_FN_14` which is not available in our HAL repo. It is provided by the `cxxsupport` in the mbed-os world.

fkjagodzinski replied:
`serial_init_direct()` is still present in our HAL API so should be tested. Adding these defines seems to be the easiest fix for that.

ithinuel commented on hal/CMakeLists.txt:
```diff
@@ -26,15 +26,14 @@ target_sources(mbed-core
         source/mbed_gpio_irq.c
+        source/mbed_gpio.c
         # source/mbed_itm_api.c
+        source/mbed_lp_ticker_api.c
```
why is this moved? it is not related to this PR.

fkjagodzinski replied:
This is an effect of sorting the list.

ithinuel commented on hal/CMakeLists.txt:
```diff
@@ -26,15 +26,14 @@ target_sources(mbed-core
-
         source/mpu/mbed_mpu_v7m.c
         source/mpu/mbed_mpu_v8m.c
+        source/static_pinmap.cpp
```
why is this added?

fkjagodzinski replied:
The commits are squashed now and I don't see a description for that, I'll check the details. This is probably required by the main.cpp.

fkjagodzinski replied:
Good catch! This change should be reverted; the test builds fine without `source/static_pinmap.cpp`. Will fix that.

fkjagodzinski replied:
Fixed via f766fcd.

ithinuel commented on hal/CMakeLists.txt:
```diff
@@ -26,15 +26,14 @@ target_sources(mbed-core
         source/mbed_flash_api.c
-        source/mbed_gpio.c
         source/mbed_gpio_irq.c
+        source/mbed_gpio.c
```
why is it moved?

fkjagodzinski replied:
I added an item to the list and then sorted the list. This is must have been in the wrong position and got sorted.

fkjagodzinski commented:
Rebased and squashed the commits. This may be considered ready if we all agree to deal with the unresolved comments in the future. :slightly_smiling_face:
